src/navigation/water_bottle_search.py

```diff
@@ -0,0 +1,426 @@
+                except SpiralEnd as spiral_error:
+                    # TODO: what to do in this case
+                    self.traj.reset()
+                    occupancy_list = None
+                except NoPath as path_error:
+                    # increment end point
+                    if self.traj.increment_point():
+                        # TODO: what to do in this case
+                        self.traj.reset()
+                    occupancy_list = None
+                if occupancy_list is None:
+                    self.star_traj = Trajectory(np.array([]))
+                else:
+                    cartesian_coords = self.ij_to_cartesian(np.array(occupancy_list))
+                    print(f"{cartesian_coords}, shape: {cartesian_coords.shape}")
+                    self.star_traj = Trajectory(
+                        np.hstack((cartesian_coords, np.zeros((cartesian_coords.shape[0], 1))))
+                    )  # current point gets set back to 0
+                    
+                    path = Path()
+                    poses = []
+                    path.header = Header()
+                    path.header.frame_id = "map"
+                    for coord in cartesian_coords:
+                        pose_stamped = PoseStamped()
+                        pose_stamped.header = Header()
+                        pose_stamped.header.frame_id = "map"
+                        point = Point(coord[0], coord[1], 0)
+                        quat = Quaternion(0,0,0,1)
+                        pose_stamped.pose = Pose(point, quat)
+                        poses.append(pose_stamped)
+                    path.poses = poses
+                    self.path_pub.publish(path)
+
+                self.time_last_updated = rospy.get_time()
+
+    def on_enter(self, context) -> None:
+        self.context = context
+        self.costmap_lock = Lock()
+        search_center = context.course.current_waypoint()
+        if not self.is_recovering:
+            self.traj = SearchTrajectory.spiral_traj(
+                context.rover.get_pose().position[0:2],
+                self.SPIRAL_COVERAGE_RADIUS,
+                self.DISTANCE_BETWEEN_SPIRALS,
+                self.SEGMENTS_PER_ROTATION,
+                search_center.tag_id,
+                True
+            )
+            self.origin = context.rover.get_pose().position[0:2]
+            print(f"ORIGIN: {self.origin}")
+            self.prev_target = None
+        self.star_traj = Trajectory(np.array([]))
+        self.time_last_updated = rospy.get_time()
+        self.costmap_listener = rospy.Subscriber("costmap", OccupancyGrid, self.costmap_callback)
+        self.path_pub = rospy.Publisher("path", Path, queue_size=10)
+
+    def on_exit(self, context) -> None:
+        self.costmap_listener.unregister()
+
+    def on_loop(self, context) -> State:
+        # continue executing the path from wherever it left off
```

Comment:
the logic for regenerating the path should probably be here in the on_loop function. You can still make it so it only goes every 0.5 seconds but ideally we have the actual logic of path generation in the state and the callback is in the context and it just sets the variable so its in line with the way the rest of the codebase works

Reply:
Resolved